flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetSingleRowJoin.scala

```diff
@@ -138,32 +146,97 @@ class DataSetSingleRowJoin(
     val condition = codeGenerator.generateExpression(joinCondition)
 
     val joinMethodBody =
-      s"""
-        |${condition.code}
-        |if (${condition.resultTerm}) {
-        |  ${conversion.code}
-        |  ${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
-        |}
-        |""".stripMargin
+      if (joinType == JoinRelType.INNER) {
+        s"""
+         |${condition.code}
+         |if (${condition.resultTerm}) {
+         |  ${conversion.code}
+         |  ${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
+         |}
+         |""".stripMargin
+    } else {
+        val singleNode =
+          if (rightIsSingle) {
+            rightNode
+          }
+          else {
+            leftNode
+          }
+
+        val notSuitedToCondition = singleNode
+          .getRowType
+          .getFieldList
+          .map(field => getRowType.getFieldNames.indexOf(field.getName))
+          .map(i => s"${conversion.resultTerm}.setField($i,null);")
+
+        if (joinType == JoinRelType.LEFT && leftIsSingle) {
+          s"""
+             |${condition.code}
+             |${conversion.code}
+             |if(!${condition.resultTerm}){
+             |${notSuitedToCondition.mkString("\n")}
+             |}
+             |if(!${condition.leftNullTerm}){
+             |${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
+             |}
+             |""".stripMargin
+        } else if (joinType == JoinRelType.RIGHT && rightIsSingle){
+          s"""
+             |${condition.code}
+             |${conversion.code}
+             |if(!${condition.resultTerm}){
+             |${notSuitedToCondition.mkString("\n")}
+             |}
+             |if(!${condition.leftNullTerm} && ${condition.resultTerm}){
+             |${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
+             |}
+             |""".stripMargin
+        } else {
+          s"""
+             |${condition.code}
+             |${conversion.code}
+             |if(!${condition.resultTerm}){
+             |${notSuitedToCondition.mkString("\n")}
+             |}
+             |${codeGenerator.collectorTerm}.collect(${conversion.resultTerm});
+             |""".stripMargin
+        }
+      }
 
     val genFunction = codeGenerator.generateFunction(
       ruleDescription,
       classOf[FlatJoinFunction[Row, Row, Row]],
       joinMethodBody,
       returnType)
 
-    if (firstIsSingle) {
-      new MapJoinRightRunner[Row, Row, Row](
-        genFunction.name,
-        genFunction.code,
-        genFunction.returnType,
-        broadcastInputSetName)
+    if (joinType == JoinRelType.RIGHT) {
```

Comment:
Do not invert the runners for `JoinRelType.RIGHT` the join input sides should be independent of the join type.